lib/KBox/src/tasks/AutoPilotTask.cpp

```diff
@@ -0,0 +1,135 @@
+/*
+  The MIT License
+
+  Copyright (c) 2016 Thomas Sarlandie thomas@sarlandie.net
+  Added by Rob McLean, March 2017
+
+  Permission is hereby granted, free of charge, to any person obtaining a copy
+  of this software and associated documentation files (the "Software"), to deal
+  in the Software without restriction, including without limitation the rights
+  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+  copies of the Software, and to permit persons to whom the Software is
+  furnished to do so, subject to the following conditions:
+
+  The above copyright notice and this permission notice shall be included in
+  all copies or substantial portions of the Software.
+
+  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+  THE SOFTWARE.
+*/
+
+/* AutoPilot functionality can exist at several levels of complexity.  In this initial iteration, the design
+ focuses on heading the boat on a defined course that is automatically set when AP "Heading" Mode is activated
+ Later on we could introduce an additional mode when a course to a waypoint is provided by a NMEA-connected device.  
+ "Dodge" mode suspends rudder actions temporarily to manually steer the boat around
+ an object or to manually put it back on course.  
+ 
+ The course calculation approach below is a simplified and adapted version of Robert Huitema's excellent FreeboardPLC_v1_2 code. RIGM
+ */
+
+#include "AutoPilotTask.h"
+#include "../pages/NavigationPage.h"
+#include "KBoxDebug.h"
+#include "../drivers/board.h"
+#include "KMessage.h"
+#include <NMEA2000.h>
+#include <N2kMessages.h>
+
+double apCurrentHeading = 0;
+double apTargetHeading = 0;
+double apTargetRudderPosition = 0;
+
+PID headingPID(&apCurrentHeading, &apTargetRudderPosition, &apTargetHeading, (double)P_Param, (double)I_Param, (double)D_Param, (int)REVERSE);
```

Comment:
In particular, I'm also not sure how I would use this code.  Does this constructor create the HeadingPid object that can be accessed in other functions on AutoPilotTask.cpp?

One of the issues I've encountered here and elsewhere is that I can only send KMessage objects from the "loop" function in any task.  If I try to do "sendMessage" from any other function nothing happens.  So I'm stuck on how to create a headingPID object, and access that object in both setup() (assuming we would like a distinct setup function) and the loop.

I'm having the same issue with the MFDTask code:  when the MFDTask::monitorN2kMessage gets an N2K message, I can extract the data, but if I try to create a MONMessage (the KMessage I'm using to send position, label and value to the display), I can't send it from within the monitorN2kMessage function.  I can only send it from the loop function in MFDTask.  Nor can I figure out how to create a MONMessage object inside the monitorN2kMessage function and then pass the object to the loop function to send the data to the display page.  Obviously there are some very different ways that c++ deals with objects than what I am used to! Any insights or practical solutions you can offer would be great.   The easiest solution here would be if I could send the Kmessage from within the function.  There's probably a switch somewhere in the visitor code structure that could let me do that.